File: csharp/Turing/Machine/IO/TuringMachineIO.cs
// <copyright file="TuringMachineIO.cs" company="Stefano Anelli">
// Copyright (c) Stefano Anelli. All rights reserved.
// </copyright>

namespace Turing.Machine.IO;

/// <summary>
/// Read the machine definition from a file.
/// </summary>
public static class TuringMachineIO
{
    /// <summary>
    /// Read a file from TOML file.
    /// </summary>
    /// <param name="textFileContent">The content of the file.</param>
    /// <returns>A turing machine setup to run the instructions described by the file.</returns>
    /// <seealso href="https://toml.io/en/"/>
    public static TuringMachine FromTomlText(string textFileContent)
    {
        var machineDescription = Tomlet.TomletMain.To<TuringMachineToml>(textFileContent);
        return new TuringMachine(
            machineDescription.States,
            machineDescription.InitialState,
            machineDescription.FinalStates,
            machineDescription.Symbols.Select(GetSymbolFromString).ToArray(),
            GetSymbolFromString(machineDescription.EmptySymbol),
            machineDescription.Transitions.Select(AsTransitionTuple).ToArray());
    }

    private static ((string State, char Symbol) From, (string State, char Symbol, TuringCursorMovement Move) To) AsTransitionTuple(TuringMachineTransitionToml transition)
        => ((transition.State, GetSymbolFromString(transition.Symbol)), (transition.NewState, GetSymbolFromString(transition.NewSymbol), GetMovementFromString(transition.Move)));

    private static char GetSymbolFromString(string value)
    {
        if (value.Length == 1)
        {
            return value[0];
        }

        throw new ArgumentException($"'{value}' is not a valid symbol");
    }

    private static TuringCursorMovement GetMovementFromString(string movement)
    {
        return new[] { TuringCursorMovement.Left, TuringCursorMovement.Right, TuringCursorMovement.None }
            .Single(enumValue => enumValue.ToString().Equals(movement, StringComparison.InvariantCultureIgnoreCase));
    }
}

File: python/turingMachineIO.py
import tomllib
from turingMachine import TuringMachine
from turingTyping import TuringState, TuringSymbol
from turingTape import TuringTapeMove

def _turingMachineFromToml(content:str) -> TuringMachine:
    toml = tomllib.loads(content)
    states:set[str] = set(toml["States"])
    initialState:str = toml["InitialState"]
    finalStates:set[str] = set(toml["FinalStates"])
    symbols:set[str] = set(toml["Symbols"])
    emptySymbol:str = toml["EmptySymbol"]
    transitions:dict[tuple[TuringState,TuringSymbol],tuple[TuringState,TuringSymbol,TuringTapeMove]] = {}

    for transition in toml["Transitions"]:
        frm = (transition["State"], transition["Symbol"])
        to = (transition["NewState"], transition["NewSymbol"], _strToTuringTapeMove(transition["Move"]))
        transitions[frm] = to

    machine = TuringMachine(states, initialState, finalStates, symbols, emptySymbol, transitions)

    return machine

def _strToTuringTapeMove(move:str) -> TuringTapeMove:
    if move == "left":
        return TuringTapeMove.Left
    elif move == "right":
        return TuringTapeMove.Right
    elif move == "none":
        return TuringTapeMove.Stay
    else:
        raise ValueError(f"Uknown move '{move}'")

def turingMachineFromFormatString(format:str, content:str) -> TuringMachine:
    if format == "toml":
        return _turingMachineFromToml(content)
    
    raise ValueError(f"Unknown format '{format}'")